src/client/opamArg.ml

```diff
@@ -782,62 +774,414 @@ let opamlist_columns =
+    let check elem =
+      check_cli_validity cli validity
+        ~cond:(fun c -> c && default != elem) elem flags
+    in
+    term_cli_check ~check Arg.(opt ?vopt kind default & doc)
+
+  let mk_opt_all ~cli validity ?section ?vopt ?(default=[])
+      flags value doc kind =
+    let doc = update_doc_w_cli doc ~cli validity in
+    let doc = Arg.info ?docs:section ~docv:value ~doc flags in
+    let check elem =
+      check_cli_validity cli validity
+        ~cond:(fun c -> c && default != elem) elem flags
+    in
+    term_cli_check ~check Arg.(opt_all ?vopt kind default & doc)
+
+  let mk_vflag ~cli ?section default flags =
+    let flags = List.map (fun (v,c,f,d) -> contented_validity v c, f, d) flags in
+    let info_flags =
+      List.map (fun (validity, flag, doc) ->
+          let doc = update_doc_w_cli doc ~cli validity in
+          validity.content, Arg.info ?docs:section flag ~doc)
+        flags
+    in
+    let check elem =
+      match
+        OpamStd.List.find_opt (fun (validity, _, _) ->
+            validity.content = elem)
+          flags
+      with
+      | Some (validity, flags, _) ->
+        check_cli_validity cli validity (elem_of_vr elem) flags
+      | None -> `Ok (elem_of_vr elem)
+    in
+    term_cli_check ~check Arg.(vflag (`Valid default) info_flags)
+
+  let mk_vflag_all ~cli ?section ?(default=[]) flags =
+    let flags = List.map (fun (v,c,f,d) -> contented_validity v c, f, d) flags in
+    let info_flags =
+      List.map (fun (validity, flag, doc) ->
+          let doc = update_doc_w_cli doc ~cli validity in
+          validity.content, Arg.info ?docs:section flag ~doc)
+        flags
+    in
+    let check selected =
+      let neww, rm =
+        List.fold_left (fun (neww,rm) elem ->
+            match OpamStd.List.find_opt (fun (validity, _, _) ->
+                validity.content = elem) flags with
+            | Some ({ removed = None ; valid = c; _ }, flags, _)
+              when cond_new cli c ->
+              (flags, c)::neww, rm
+            | Some ({ removed = Some (removal, instead); _ }, flags, _)
+              when cond_removed cli removal ->
+              neww, (flags, (removal, instead))::rm
+            | _ -> neww,rm) ([],[]) selected
+      in
+      match neww, rm with
+      | [], [] -> `Ok (List.map elem_of_vr selected)
+      | [flags, c], [] ->
+        error_new_flag cli c flags
+      | [], [flags, (c, instead)] ->
+        error_rm_flag cli c instead flags
+      | _::_, []->
+        let options, clis = List.split neww in
+        let msg =
+          Printf.sprintf
+            "%s are only available since respectively versions %s of the opam \
+             CLI, but version %s has been requested."
+            (OpamStd.Format.pretty_list (List.map get_long_form options) )
+            (OpamStd.Format.pretty_list
+               (List.map OpamCLIVersion.to_string clis))
+            (OpamCLIVersion.to_string cli)
+        in
```

Comment:
I'd alter this message to look for the maximum CLI and instead display:
```
            "%s can only be used with at least version %s of the opam \
             CLI, but version %s has been requested."
```
i.e. tell the user what they need for this specific combination, rather than for each individual flag (especially as if we display "respectively", it's actually nicer to group them by version, which starts to become a huge function for an error message!)

Reply:
(UI, and good error message is always a long part :))